File: 0278-first-bad-version/0278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int firstBadProduct=-1;
        int start=1;
        int end=n;
        while(start<=end){
            int mid=start+(end-start)/2;
            if(isBadVersion(mid)){
                firstBadProduct=mid;
                end=mid-1;
            }
            else start=mid+1;
        }
        return firstBadProduct;
    }
}

File: 0278-first-bad-version/0278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        firstBadProduct=float('inf')
        left=1
        right=n
        while left<=right:
            middle=left+(right-left)//2
            if isBadVersion(middle):
                firstBadProduct=middle
                right=middle-1
            else:
                left=middle+1
        return firstBadProduct
        
        